lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/DriverStatusManager.java

```diff
@@ -143,113 +160,121 @@ public synchronized void onError(final Throwable exception) {
       this.clock.stop(exception);
       this.setStatus(DriverStatus.FAILING);
     }
-    LOG.exiting(DriverStatusManager.class.getCanonicalName(), "onError", new Object[]{exception});
+
+    LOG.exiting(CLASS_NAME, "onError", exception);
   }
 
   /**
    * Perform a clean shutdown of the Driver.
    */
-  @SuppressWarnings("checkstyle:constructorwithoutparams") // Exception() here captures the callstack
   public synchronized void onComplete() {
-    LOG.entering(DriverStatusManager.class.getCanonicalName(), "onComplete");
+
+    LOG.entering(CLASS_NAME, "onComplete");
+
     if (this.isShuttingDownOrFailing()) {
-      LOG.log(Level.WARNING, "Ignoring second call to onComplete()");
+      LOG.log(Level.WARNING, "Ignoring second call to onComplete()",
+          new Exception("Dummy exception to get the call stack"));
     } else {
       LOG.log(Level.INFO, "Clean shutdown of the Driver.");
       if (LOG.isLoggable(Level.FINEST)) {
-        LOG.log(Level.FINEST, "Callstack: ", new Exception());
+        LOG.log(Level.FINEST, "Call stack: ",
+            new Exception("Dummy exception to get the call stack"));
       }
       this.clock.close();
       this.setStatus(DriverStatus.SHUTTING_DOWN);
     }
-    LOG.exiting(DriverStatusManager.class.getCanonicalName(), "onComplete");
 
+    LOG.exiting(CLASS_NAME, "onComplete");
+  }
+
+  /**
+   * Sends the final message to the client. This is used by DriverRuntimeStopHandler.onNext().
+   * @param exception Exception that caused the job to end (optional).
+   */
+  public synchronized void onRuntimeStop(final Optional<Throwable> exception) {
+    this.sendJobEndingMessageToClient(exception);
   }
 
   /**
    * Sends the final message to the Driver. This is used by DriverRuntimeStopHandler.onNext().
-   *
-   * @param exception
+   * @param exception Exception that caused the job to end (can be absent).
+   * @deprecated [REEF-1548] Do not use DriverStatusManager as a proxy to the job client.
+   * After release 0.16, make this method private and use it inside onRuntimeStop() method instead.
```

Comment:
[REEF-1548](https://issues.apache.org/jira/browse/REEF-1548) *is* the relevant issue